tests/admin_utils/tests.py

```diff
@@ -102,6 +102,52 @@ def test_relation_on_abstract(self):
 class UtilsTests(SimpleTestCase):
     empty_value = "-empty-"
 
+    def _empty_display_for_field(self, value):
+        display_value = display_for_field(value, models.CharField(), self.empty_value)
+        self.assertEqual(display_value, self.empty_value)
+
+        display_value = display_for_field(
+            value,
+            models.CharField(choices=((value, "test_empty"),)),
+            self.empty_value,
+        )
+        self.assertEqual(display_value, "test_empty")
+
+        display_value = display_for_field(value, models.DateField(), self.empty_value)
+        self.assertEqual(display_value, self.empty_value)
+
+        display_value = display_for_field(value, models.TimeField(), self.empty_value)
+        self.assertEqual(display_value, self.empty_value)
+
+        display_value = display_for_field(
+            value, models.BooleanField(null=True), self.empty_value
+        )
+        expected = '<img src="%sadmin/img/icon-unknown.svg" alt="%s" />' % (
+            settings.STATIC_URL,
+            value,
+        )
+        self.assertHTMLEqual(display_value, expected)
+
+        display_value = display_for_field(
+            value, models.DecimalField(), self.empty_value
+        )
+        self.assertEqual(display_value, self.empty_value)
+
+        display_value = display_for_field(value, models.FloatField(), self.empty_value)
+        self.assertEqual(display_value, self.empty_value)
+
+        display_value = display_for_field(value, models.JSONField(), self.empty_value)
+        self.assertEqual(display_value, self.empty_value)
+
+    def test_empty_display_for_field(self):
+        """
+        Regression test for #12550: display_for_field should handle None value.
+        Regression test for #28404: Django admin empty_value/empty_value_display doesn't
+                                    check for empty strings
+        """
+        for field_value in [None, ""]:
+            self._empty_display_for_field(field_value)
```

Comment:
I'd use `subTest()` instead to get a better test coverage and make it more readable.
- 1st commit to split `test_null_display_for_field` into three tests:
```python
    def test_empty_value_display_for_field(self):
        tests = [
            models.CharField(),
            models.DateField(),
            models.DecimalField(),
            models.FloatField(),
            models.JSONField(),
            models.TimeField(),
        ]
        for model_field in tests:
            with self.subTest(model_field=model_field):
                display_value = display_for_field(
                    None, model_field, self.empty_value
                )   
                self.assertEqual(display_value, self.empty_value)
                
    def test_empty_value_display_choices(self):
        model_field = models.CharField(choices=((None, "test_none"),))
        display_value = display_for_field(None, model_field, self.empty_value)
        self.assertEqual(display_value, "test_none")
        
    def test_empty_value_display_booleanfield(self):
        model_field = models.BooleanField(null=True)
        display_value = display_for_field(None, model_field, self.empty_value)
        expected = (
            f'<img src="{settings.STATIC_URL}admin/img/icon-unknown.svg" alt="None" />'
        )   
        self.assertHTMLEqual(display_value, expected)

```
- 2nd commit with your patch and the following test changes:
```diff
diff --git a/tests/admin_utils/tests.py b/tests/admin_utils/tests.py
index ecf9ee9795..393770bd2d 100644
--- a/tests/admin_utils/tests.py
+++ b/tests/admin_utils/tests.py
@@ -159,11 +159,12 @@ class UtilsTests(SimpleTestCase):
             models.TimeField(),
         ]
         for model_field in tests:
-            with self.subTest(model_field=model_field):
-                display_value = display_for_field(
-                    None, model_field, self.empty_value
-                )
-                self.assertEqual(display_value, self.empty_value)
+            for value in model_field.empty_values:
+                with self.subTest(model_field=model_field, empty_value=value):
+                    display_value = display_for_field(
+                        value, model_field, self.empty_value
+                    )
+                    self.assertEqual(display_value, self.empty_value)
 
     def test_empty_value_display_choices(self):
         model_field = models.CharField(choices=((None, "test_none"),))
```

Reply:
Didn't know about `subTest()`. This looks way better indeed.